pkg/capacityscheduling/capacity_scheduling.go

```diff
@@ -0,0 +1,869 @@
+		var ok bool
+		args, ok := obj.(*config.CapacitySchedulingArgs)
+		if !ok {
+			return nil, fmt.Errorf("want args to be of type CapacitySchedulingArgs, got %T", obj)
+		}
+		kubeConfigPath = args.KubeConfigPath
+	}
+
+	c := &CapacityScheduling{
+		frameworkHandle: handle,
+		elasticQuotas:   make(map[string]*ElasticQuotaInfo),
+		pdbLister:       getPDBLister(handle.SharedInformerFactory()),
+		lock:            &sync.RWMutex{},
+		assumedPods:     make(map[string]bool),
+	}
+
+	restConfig, err := clientcmd.BuildConfigFromFlags("", kubeConfigPath)
+	if err != nil {
+		return nil, err
+	}
+	client, err := versioned.NewForConfig(restConfig)
+	if err != nil {
+		return nil, err
+	}
+	externalSharedInformerFactory := externalversions.NewSharedInformerFactory(client, 0)
+	elasticQuotaInformer := externalSharedInformerFactory.Scheduling().V1alpha1().ElasticQuotas().Informer()
+	elasticQuotaInformer.AddEventHandler(
+		cache.FilteringResourceEventHandler{
+			FilterFunc: func(obj interface{}) bool {
+				switch t := obj.(type) {
+				case *v1alpha1.ElasticQuota:
+					return true
+				case cache.DeletedFinalStateUnknown:
+					if _, ok := t.Obj.(*v1alpha1.ElasticQuota); ok {
+						return true
+					}
+					utilruntime.HandleError(fmt.Errorf("cannot convert to *v1.Pod: %v", obj))
+					return false
+				default:
+					utilruntime.HandleError(fmt.Errorf("unable to handle object in %T", obj))
+					return false
+				}
+			},
+			Handler: cache.ResourceEventHandlerFuncs{
+				AddFunc:    c.addElasticQuota,
+				UpdateFunc: c.updateElasticQuota,
+				DeleteFunc: c.deleteElasticQuota,
+			},
+		})
+
+	go elasticQuotaInformer.Run(nil)
+	if !cache.WaitForCacheSync(nil, elasticQuotaInformer.HasSynced) {
+		klog.Errorf("Timed out waiting for caches to sync")
+	}
+
+	podInformer := handle.SharedInformerFactory().Core().V1().Pods().Informer()
+	podInformer.AddEventHandler(
+		cache.FilteringResourceEventHandler{
+			FilterFunc: func(obj interface{}) bool {
+				switch t := obj.(type) {
+				case *v1.Pod:
+					return assignedPod(t)
+				case cache.DeletedFinalStateUnknown:
+					if pod, ok := t.Obj.(*v1.Pod); ok {
+						return assignedPod(pod)
+					}
+					return false
+				default:
+					return false
+				}
+			},
+			Handler: cache.ResourceEventHandlerFuncs{
+				AddFunc:    c.addPod,
+				DeleteFunc: c.deletePod,
+			},
+		},
+	)
+
+	return c, nil
+}
+
+// PreFilter performs the following validations.
+func (c *CapacityScheduling) PreFilter(ctx context.Context, state *framework.CycleState, pod *v1.Pod) *framework.Status {
+	prefilterStatue := computePodResourceRequest(pod)
+	snapshotState := c.createSnapshot()
+	elasticQuotaInfo := snapshotState.elasticQuotas[pod.Namespace]
+
+	if !elasticQuotaInfo.lessThanMax(prefilterStatue.Resource) {
+		return framework.NewStatus(framework.Error, "MoreThanMax")
+	}
+	state.Write(preFilterStateKey, prefilterStatue)
+	state.Write(SnapshotStateKey, snapshotState)
+	return framework.NewStatus(framework.Success, "")
```

Comment:
Where is the logic of verifying the pods **needs** a preemption? I assume it should be in Filter():

- including the incoming pod's req, current EQ's `used` <= `min`, then for sure return `Success`.
- including the incoming pod's req, `min` < current EQ's `used` <= `max` and sum(eq's `used`) > sum(eq's `min`), so we should return `Unschedulable`

Reply:
bq. including the incoming pod's req, min < current EQ's used <= max and sum(eq's used) > sum(eq's min), so we should return Unschedulable.

I do not understand why we should return Unschedulable here. As long as the current EQ's total used is smaller than the max, 

- If the cluster still has the available resource, it could be assigned

- Otherwise, based on the priority, the preemption could happen in the same namespace(Line 444)